types/currency.go

```diff
@@ -151,15 +152,25 @@ func (x Currency) Sqrt() (c Currency) {
 func (x Currency) Sub(y Currency) (c Currency) {
 	if x.Cmp(y) < 0 {
 		c = x
-		if build.DEBUG {
-			panic(ErrNegativeCurrency)
-		}
+		build.Critical(ErrNegativeCurrency)
 	} else {
 		c.i.Sub(&x.i, &y.i)
 	}
 	return
 }
 
+// Uint64 converts a Currency to a uint64. An error is returned because this
+// function is sometimes called on values that can be determined by users -
+// rather than have all user-facing points do input checking, the input
+// checking should happen at the base type. This minimizes the chances of a
+// rogue user causing a build.Critical to be triggered.
+func (c Currency) Uint64() (u uint64, err error) {
+	if c.Cmp(NewCurrency64(math.MaxUint64)) > 0 {
+		return 0, ErrUint64Overflow
+	}
+	return uint64(c.Big().Int64()), nil
+}
```

Comment:
and, my implementation is incorrect anyway as an int64 isn't as large as a uint64, so I definitely missed an edge case here during testing.